dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusC.java

```diff
@@ -577,4 +724,27 @@ private Expression assignToReturnRegister(Register register, Expression value) {
         }
         return value;
     }
+
+    private void addMemoryObjectTags(Event event, Expression address) {
+        if (address instanceof Register reg) {
+            MemoryObject object = programBuilder.getLocFromReg(reg);
+            if (object != null) {
+                event.addTags(object.getFeatureTags());
+            }
+        } else {
+            for (Expression add : address.getOperands()) {
+                addMemoryObjectTags(event, add);
+            }
+        }
+    }
```

Comment:
I don't really understand what this method is about

Reply:
From what I understand, this code tags events that access specially-marked locations. I.e., if location `x` is marked, then `load(&x)` gets marked as well. The issue is that a load may refer to `x` like this: `r <- &x; load(r)` in which case the accessed location is obscured by the register (and this is I think also the reason for why `getLocFromReg` exists).

Now, this approach is only feasible if there are no unclear aliasing relations, i.e., it is statically known which memory event accesses which location. This might be viable for litmus code but I don't think we can extend this to general OpenCL code.
A workaround might be to compute the tags within the CAT by observing the init event that belongs to the address:
```
let myDynamicTag = [range([IW & myStaticTag]; loc)] 
// myDynamicTag relates events to itself that access an address whose init event is marked myStaticTag
```
This way only only init events need to get static tagging.
In case there are no init events, e.g. due to dynamic allocation, then there is also no tagging. I don't know if OpenCL allows to dynamically allocate tagged memory.

Reply:
This seems to be analogous to what we have in real spirv code. There were some tags (related to storage classes IIRC) that were actually related to the memory object being accessed. If this is the case, we should probably handle this in the same way it is currently handled by the spirv code.

Reply:
I took a quick glance at the SPIRV code and it seems that the parser relies on a sort of explicit tagging of pointers used in memory accesses. Put differently, in SPIRV pointers are explicitly tagged rather than implicitly by the memory they point to. This also suggests that you could technically have two same-address pointers but with different storage classes (might be UB though?). In other words, storage classes are not per address but per access (like access types) and so a memory access has the same storage class independent of the actually accessed address.
I think this makes sense, because the compiler generates different assembly based on the storage classes and so it must statically be able to infer them.

I don't know if the tagging @tonghaining implemented here is supposed to represent the same kind of thing.
Basically the question is: is the tagging supposed to model different assembly generated for different code (-> static tagging via tagged pointers) or is it supposed to model the hardware doing different things (for the same instruction) depending on the internal memory layout (-> dynamic tagging based on accessed address)?

Reply:
Ok, then both memory and pointers have tags and these are different. Memory accesses inherit the tags of their pointers and so pointers need to carry this extra information like they do for SPIRV.
Interestingly, this tag inheritance cannot always be done statically because of [genericAddressSpace](https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/genericAddressSpace.html).

Now, assuming that an object in space X is only accessible with pointers to space X (and otherwise it is UB), there is really not much difference between computing the tagging based on the pointer type vs. based on the actually accessed address (since they will match). The former can be done statically in many cases except for "genericAddressSpace" while the latter is always fully dynamic.
I think the dynamic tagging is easier to implement in CAT as I described above and does not require special pointer types(*). On the other hand, we cannot detect UB based on mismatching tagging between pointer and memory (constructing such a mismatch in real OpenCL code is likely impossible anyways).

(*) We will eventually want to have proper pointer types with address spaces, but this is a more invasive and complex change.